contributing/Debugging.md

```diff
@@ -0,0 +1,168 @@
+
+The build should create debugging run configurations for you [automatically](../build-logic/idea/src/main/kotlin/gradlebuild.ide.gradle.kts#L82), but you'll need to make sure to enable automatic restart of the debugging session once the debuggee is gone.
+Unfortunately, setting this up [is not automated yet](https://github.com/JetBrains/gradle-idea-ext-plugin/issues/84).
+
+![](./images/auto-restart-debugger.png)
+
+Before you start debugging, start the "Daemon debug" or "Debug Launcher" configuration. 
+It should be automatically re-used for subsequent test runs.
+
+Be careful when you stop the test: you should stop the test, not the debugger.
+
+### Gradle test executors
+
+There are [several executors](../testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/GradleContextualExecuter.java) available for running integration tests configured in Gradle.
+
+Most of them are forking, and you should be able to debug a test with any of them by running the test in the debug mode.
+The test detects that it is running in debug mode and provides all necessary information to the forked daemon to connect to the debugger.
+
+There is also an embedded executor that runs the test in the same JVM as the test runner.
+While it's convenient for debugging and sometimes may be faster, it's not recommended to use it because it may behave differently in some corner cases due to classloader issues.
+
+In test code, you can access the current executer via `executer` and modify its options.
+
+### Gradle debug options
+
+You may also refer to these resources about debugging Gradle build in general:
+* [Gradle Troubleshooting](https://docs.gradle.org/current/userguide/troubleshooting.html)
+* [Debugging options](https://docs.gradle.org/current/userguide/command_line_interface.html#sec:command_line_debugging) 
+
+The [blog series](https://blog.gradle.org/how-gradle-works-1) about how Gradle works can help understand why multiple JVMs are involved in the build process.
+
+## Debugging daemon
+
+Just starting the test in debug mode in IDE with the "Debug Daemon" configuration running should do the trick.
+
+You can also explicitly enable debugging for the daemon by setting `debugDaemon` gradle property to any value or changing the test code with `executer.startBuildProcessInDebugger` method.
+
+## Debugging launcher
+
+You can enable debugging for the launcher by setting `debugLauncher` gradle property to any value or changing the test code with `executer.startLauncherInDebugger` method.
+
+Note that by default, port 5006 is used for debugging the launcher, so you'll also need to start the "Debug Launcher" configuration.
+
+## Debugging wrapper
+
+The main obstacle to wrapper debugging is the minification of the wrapper jar. 
+To mitigate this, minifying is disabled when the `debugLauncher` property is present.
+Note that means that it wouldn't be 100% the same as in production and may behave differently in corner cases.
```

Comment:
```suggestion
Keep in mind that this may not be exactly the same as the production environment and could behave differently in some corner cases.
```